usr/src/lib/smbclnt/libfknsmb/common/fake_stream.c

```diff
@@ -0,0 +1,1834 @@
+		crfree(dbp->db_credp);
+		dbp->db_credp = NULL;
+	}
+	dbp->db_cpid = -1;
+	dbp->db_struioflag = 0;
+	dbp->db_struioun.cksum.flags = 0;
+
+	kmem_cache_free(dbp->db_cache, dbp);
+}
+
+/*ARGSUSED*/
+static void
+frnop_func(void *arg)
+{
+}
+
+/*
+ * Generic esballoc used to implement the four flavors: [d]esballoc[a].
+ */
+static mblk_t *
+gesballoc(unsigned char *base, size_t size, uint32_t db_rtfu, frtn_t *frp,
+	void (*lastfree)(mblk_t *, dblk_t *), int kmflags)
+{
+	dblk_t *dbp;
+	mblk_t *mp;
+
+	ASSERT(base != NULL && frp != NULL);
+
+	if ((dbp = kmem_cache_alloc(dblk_esb_cache, kmflags)) == NULL) {
+		mp = NULL;
+		goto out;
+	}
+
+	mp = dbp->db_mblk;
+	dbp->db_base = base;
+	dbp->db_lim = base + size;
+	dbp->db_free = dbp->db_lastfree = lastfree;
+	dbp->db_frtnp = frp;
+	DBLK_RTFU_WORD(dbp) = db_rtfu;
+	mp->b_next = mp->b_prev = mp->b_cont = NULL;
+	mp->b_rptr = mp->b_wptr = base;
+	mp->b_queue = NULL;
+	MBLK_BAND_FLAG_WORD(mp) = 0;
+
+out:
+	FTRACE_1("gesballoc(): mp=0x%lx", (uintptr_t)mp);
+	return (mp);
+}
+
+/*ARGSUSED*/
+mblk_t *
+esballoc(unsigned char *base, size_t size, uint_t pri, frtn_t *frp)
+{
+	mblk_t *mp;
+
+	/*
+	 * Note that this is structured to allow the common case (i.e.
+	 * STREAMS flowtracing disabled) to call gesballoc() with tail
+	 * call optimization.
+	 */
+	if (!str_ftnever) {
+		mp = gesballoc(base, size, DBLK_RTFU(1, M_DATA, 0, 0),
+		    frp, freebs_enqueue, KM_NOSLEEP);
+
+		if (mp != NULL)
+			STR_FTALLOC(&DB_FTHDR(mp), FTEV_ESBALLOC, size);
+		return (mp);
+	}
+
+	return (gesballoc(base, size, DBLK_RTFU(1, M_DATA, 0, 0),
+	    frp, freebs_enqueue, KM_NOSLEEP));
+}
+
+/*
+ * Same as esballoc() but sleeps waiting for memory.
+ */
+/*ARGSUSED*/
+mblk_t *
+esballoc_wait(unsigned char *base, size_t size, uint_t pri, frtn_t *frp)
+{
+	mblk_t *mp;
+
+	/*
+	 * Note that this is structured to allow the common case (i.e.
+	 * STREAMS flowtracing disabled) to call gesballoc() with tail
+	 * call optimization.
+	 */
+	if (!str_ftnever) {
+		mp = gesballoc(base, size, DBLK_RTFU(1, M_DATA, 0, 0),
+		    frp, freebs_enqueue, KM_SLEEP);
+
+		STR_FTALLOC(&DB_FTHDR(mp), FTEV_ESBALLOC, size);
+		return (mp);
+	}
+
+	return (gesballoc(base, size, DBLK_RTFU(1, M_DATA, 0, 0),
+	    frp, freebs_enqueue, KM_SLEEP));
+}
+
+/*ARGSUSED*/
+mblk_t *
+desballoc(unsigned char *base, size_t size, uint_t pri, frtn_t *frp)
+{
+	mblk_t *mp;
+
+	/*
+	 * Note that this is structured to allow the common case (i.e.
+	 * STREAMS flowtracing disabled) to call gesballoc() with tail
+	 * call optimization.
+	 */
+	if (!str_ftnever) {
+		mp = gesballoc(base, size, DBLK_RTFU(1, M_DATA, 0, 0),
+		    frp, dblk_lastfree_desb, KM_NOSLEEP);
+
+		if (mp != NULL)
+			STR_FTALLOC(&DB_FTHDR(mp), FTEV_DESBALLOC, size);
+		return (mp);
+	}
+
+	return (gesballoc(base, size, DBLK_RTFU(1, M_DATA, 0, 0),
+	    frp, dblk_lastfree_desb, KM_NOSLEEP));
+}
+
+/*ARGSUSED*/
+mblk_t *
+esballoca(unsigned char *base, size_t size, uint_t pri, frtn_t *frp)
```

Comment:
I'd be fairly stunned if any of your user space code or even your in-kernel code ever neeed desballoc or esballoca.  These are for device drivers that want to loan up premapped buffers, and should not be called by anything that isn't a hardware driver I believe.